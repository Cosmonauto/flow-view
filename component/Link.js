const bindme = require('bindme')
const pdsp = require('pdsp')
const staticProps = require('static-props')

const Component = require('./Component')

/**
 * A Link connects a node output to another node input.
 */

class FlowViewLink extends Component {
  constructor (canvas, dispatch, container) {
    super(canvas, dispatch, container)

    // DOM Elements.
    // =================================================================

    const line = this.createElementNS('path')
    const start = this.createElementNS('circle')
    const end = this.createElementNS('circle')

    // Static attributes.
    // =================================================================

    staticProps(this)({
      end,
      line,
      id: () => container.getAttribute('id'),
      start,
      theme: () => canvas.theme.link
    })

    // Event bindings.
    // =================================================================

    bindme(this,
      'onClick',
      'onDblclick'
    )

    container.addEventListener('click', this.onClick)
    container.addEventListener('dblclick', this.onDblclick)
    container.addEventListener('mousedown', this.dropEvent)
  }

  onClick (event) {
    pdsp(event)

    const { id, dispatch } = this

    dispatch('selectLink', id)
  }

  onDblclick (event) {
    pdsp(event)

    const { id, dispatch } = this

    dispatch('deleteLink', id)
  }

  render (state) {
    const {
      canvas,
      line,
      start,
      end
    } = this

    const {
      // TODO graph,
      endX,
      endY,
      selected,
      startX,
      startY
    } = state

    const pinSize = canvas.theme.pin.size

    const {
      baseColor,
      highlightColor,
      width
    } = this.theme

    // TODO const {from, to} = graph
    // TODO const connected = (typeof from !== 'undefined') && (typeof to !== 'undefined')

    // Changed properties.
    // =================================================================

    const endXChanged = (endX !== this.endX)
    const endYChanged = (endY !== this.endY)
    const pinSizeChanged = (pinSize !== this.pinSize)
    const selectedChanged = (selected !== this.selected)
    const startXChanged = (startX !== this.startX)
    const startYChanged = (startY !== this.startY)
    const widthChanged = (width !== this.width)

    // Link width.
    // =================================================================

    if (widthChanged) {
      this.width = width

      line.setAttribute('stroke-width', width)
    }

    // Link color.
    // =================================================================

    if (selectedChanged) {
      this.selected = selected

      if (selected) {
        line.setAttribute('stroke', highlightColor)
        start.setAttribute('fill', highlightColor)
        end.setAttribute('fill', highlightColor)
      } else {
        line.setAttribute('stroke', baseColor)
        start.setAttribute('fill', baseColor)
        end.setAttribute('fill', baseColor)
      }
    }

    // Start and end position.
    // =================================================================

    if (startXChanged) {
      this.startX = startX
      start.setAttribute('cx', startX)
    }

    if (startYChanged) {
      this.startY = startY
      start.setAttribute('cy', startY)
    }

    if (endXChanged) {
      this.endX = endX
      end.setAttribute('cx', endX)
    }

    if (endYChanged) {
      this.endY = endY
      end.setAttribute('cy', endY)
    }

    if (startXChanged || startYChanged || endXChanged || endYChanged) {
      // TODO Code for bezier
      // const midPointY = (startY + endY) / 2
      //
      // const controlPointX1 = startX
      // const controlPointY1 = midPointY
      // const controlPointX2 = endX
      // const controlPointY2 = midPointY
      //
      // `M ${startX} ${startY} C ${controlPointX1} ${controlPointY1}, ${controlPointX2} ${controlPointY2} ,${endX} ${endY}`

      line.setAttribute('d', `M ${startX} ${startY} L ${endX} ${endY}`)
    }

    // Start and end size.
    // =================================================================

    if (pinSizeChanged) {
      this.pinSize = pinSize
      start.setAttribute('r', pinSize / 2)
      end.setAttribute('r', pinSize / 2)
    }
  }
}

module.exports = exports.default = FlowViewLink
