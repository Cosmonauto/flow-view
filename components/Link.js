(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', 'react', '../utils/ignoreEvent', './theme'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('react'), require('../utils/ignoreEvent'), require('./theme'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.react, global.ignoreEvent, global.theme);
    global.Link = mod.exports;
  }
})(this, function (module, exports, _react, _ignoreEvent, _theme) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _ignoreEvent2 = _interopRequireDefault(_ignoreEvent);

  var _theme2 = _interopRequireDefault(_theme);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var Link = function (_Component) {
    _inherits(Link, _Component);

    function Link() {
      _classCallCheck(this, Link);

      return _possibleConstructorReturn(this, (Link.__proto__ || Object.getPrototypeOf(Link)).apply(this, arguments));
    }

    _createClass(Link, [{
      key: 'render',
      value: function render() {
        var _props = this.props,
            id = _props.id,
            deleteLink = _props.deleteLink,
            from = _props.from,
            onCreateLink = _props.onCreateLink,
            startDraggingLinkTarget = _props.startDraggingLinkTarget,
            selected = _props.selected,
            selectLink = _props.selectLink,
            sourceSelected = _props.sourceSelected,
            targetSelected = _props.targetSelected,
            theme = _props.theme,
            to = _props.to,
            x1 = _props.x1,
            y1 = _props.y1,
            x2 = _props.x2,
            y2 = _props.y2;
        var darkPrimaryColor = theme.darkPrimaryColor,
            primaryColor = theme.primaryColor,
            linkColor = theme.linkColor,
            lineWidth = theme.lineWidth,
            pinSize = theme.pinSize;


        var onSourceMouseDown = function onSourceMouseDown(e) {
          e.preventDefault();
          e.stopPropagation();

          onCreateLink({ from: from, to: null });
        };

        var onTargetMouseDown = function onTargetMouseDown(e) {
          e.preventDefault();
          e.stopPropagation();

          startDraggingLinkTarget(id);
        };

        var startX = x1 + pinSize / 2;
        var startY = y1 + pinSize / 2;
        var endX = x2 + pinSize / 2;
        var endY = y2 + pinSize / 2;

        var midPointY = (startY + endY) / 2;

        var controlPointX1 = startX;
        var controlPointY1 = to ? midPointY : startY;
        var controlPointX2 = endX;
        var controlPointY2 = to ? midPointY : endY;

        return _react2.default.createElement(
          'g',
          {
            onClick: _ignoreEvent2.default,
            onDoubleClick: _ignoreEvent2.default
          },
          _react2.default.createElement('path', {
            d: 'M ' + startX + ' ' + startY + ' C ' + controlPointX1 + ' ' + controlPointY1 + ', ' + controlPointX2 + ' ' + controlPointY2 + ' ,' + endX + ' ' + endY,
            fill: 'transparent',
            onMouseDown: function onMouseDown() {
              if (selected) deleteLink(id);
            },
            onMouseUp: selectLink,
            stroke: selected ? primaryColor : linkColor,
            strokeWidth: lineWidth
          }),
          _react2.default.createElement('rect', {
            fill: selected || sourceSelected ? darkPrimaryColor : linkColor,
            height: pinSize,
            onMouseDown: onSourceMouseDown,
            width: pinSize,
            x: x1,
            y: y1
          }),
          to ? _react2.default.createElement('rect', {
            fill: selected || targetSelected ? darkPrimaryColor : linkColor,
            height: pinSize,
            onMouseDown: onTargetMouseDown,
            width: pinSize,
            x: x2,
            y: y2
          }) : null
        );
      }
    }]);

    return Link;
  }(_react.Component);

  Link.propTypes = {
    deleteLink: _react.PropTypes.func.isRequired,
    id: _react.PropTypes.string,
    from: _react.PropTypes.array,
    onCreateLink: _react.PropTypes.func.isRequired,
    startDraggingLinkTarget: _react.PropTypes.func.isRequired,
    pinSize: _react.PropTypes.number.isRequired,
    selected: _react.PropTypes.bool.isRequired,
    selectLink: _react.PropTypes.func.isRequired,
    sourceSelected: _react.PropTypes.bool.isRequired,
    targetSelected: _react.PropTypes.bool.isRequired,
    theme: _theme2.default.propTypes,
    to: _react.PropTypes.array,
    x1: _react.PropTypes.number.isRequired,
    y1: _react.PropTypes.number.isRequired,
    x2: _react.PropTypes.number.isRequired,
    y2: _react.PropTypes.number.isRequired
  };

  Link.defaultProps = {
    deleteLink: Function.prototype,
    onCreateLink: Function.prototype,
    startDraggingLinkTarget: Function.prototype,
    selected: false,
    selectLink: Function.prototype,
    sourceSelected: false,
    targetSelected: false,
    theme: _theme2.default.defaultProps
  };

  exports.default = Link;
  module.exports = exports['default'];
});