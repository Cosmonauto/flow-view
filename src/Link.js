
function Link (canvas, view, key) {
  var draw = canvas.draw

  var theme = canvas.theme

  var strokeDasharray = theme.strokeDasharray,
      strokeLine      = theme.strokeLine

  var from = canvas.box[view.from[0]],
      to   = canvas.box[view.to[0]]

  var start = from.outs[view.from[1]],
      end   = to.ins[view.to[1]]

  Object.defineProperty(this, 'x1', { get: function () { return start.center.absolute.x } })
  Object.defineProperty(this, 'y1', { get: function () { return start.center.absolute.y } })
  Object.defineProperty(this, 'x2', { get: function () { return end.center.absolute.x } })
  Object.defineProperty(this, 'y2', { get: function () { return end.center.absolute.y } })

  var line = this.line = draw.line(this.x1, this.y1, this.x2, this.y2)
                             .stroke(strokeLine)
                             .attr('stroke-dasharray', strokeDasharray)

  end.link = this
  start.link[key] = this

  function remove () {
    end.link = null
    delete start.link[key]
    delete canvas.view.link[key]
    line.remove()
  }

  function deselectLine () {
    line.off('click')
        .attr('stroke-dasharray', strokeDasharray)
  }

  function selectLine () {
    line.on('click', remove)
        .attr('stroke-dasharray', null)
  }

  line.on('mouseover', selectLine)
  line.on('mouseout', deselectLine)
}

module.exports = Link

