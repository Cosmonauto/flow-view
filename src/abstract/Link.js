import from './CanvasItem'

class FlowViewLink extends CanvasItem {
  constructor (canvas) {
    super(canvas)
  }
}

export default FlowViewLink
