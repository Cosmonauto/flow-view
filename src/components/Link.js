// @flow
import React from 'react'

import bindme from 'bindme'

import ignoreEvent from '../utils/ignoreEvent'
import { defaultTheme, Theme } from './theme'
import {
  Id,
  NodeIdAndPosition
} from './types'

export default class Link extends React.PureComponent {
  props: {
    createLink: ({ from: NodeIdAndPosition, to?: NodeIdAndPosition }) => Id,
    deleteLink: (string) => void,
    id: string,
    from: NodeIdAndPosition,
    startDraggingLinkTarget: (Id) => void,
    pinSize: number,
    selected: boolean,
    selectLink: (MouseEvent) => void,
    sourceSelected: boolean,
    targetSelected: boolean,
    theme: Theme,
    to: NodeIdAndPosition,
    x1: number,
    y1: number,
    x2: number,
    y2: number
  }

  static defaultProps = {
    createLink: Function.prototype,
    deleteLink: Function.prototype,
    startDraggingLinkTarget: Function.prototype,
    selected: false,
    selectLink: Function.prototype,
    sourceSelected: false,
    targetSelected: false,
    theme: defaultTheme
  }

  constructor () {
    bindme(super(),
      'onPathMouseDown',
      'onSourceMouseDown',
      'onTargetMouseDown'
    )
  }

  onPathMouseDown (event: MouseEvent): void {
    event.preventDefault()
    event.stopPropagation()

    const {
      id,
      deleteLink,
      selected
    } = this.props

    if (selected) deleteLink(id)
  }

  onSourceMouseDown (event: MouseEvent): void {
    event.preventDefault()
    event.stopPropagation()

    const {
      from,
      createLink
    } = this.props

    createLink({ from })
  }

  onTargetMouseDown (event: MouseEvent): void {
    event.preventDefault()
    event.stopPropagation()

    const {
      id,
      startDraggingLinkTarget
    } = this.props

    startDraggingLinkTarget(id)
  }

  render () {
    const {
      selected,
      selectLink,
      sourceSelected,
      targetSelected,
      theme,
      to,
      x1,
      y1,
      x2,
      y2
    } = this.props

    const {
      darkPrimaryColor,
      primaryColor,
      linkColor,
      lineWidth,
      pinSize
    } = theme

    const startX = x1 + (pinSize / 2)
    const startY = y1 + (pinSize / 2)
    const endX = x2 + (pinSize / 2)
    const endY = y2 + (pinSize / 2)

    const midPointY = (startY + endY) / 2

    const controlPointX1 = startX
    const controlPointY1 = to ? midPointY : startY
    const controlPointX2 = endX
    const controlPointY2 = to ? midPointY : endY

    return (
      <g
        onClick={ignoreEvent}
        onDoubleClick={ignoreEvent}
      >
        <path
          d={startY <= endY ? `M ${startX} ${startY} C ${controlPointX1} ${controlPointY1}, ${controlPointX2} ${controlPointY2} ,${endX} ${endY}` : `M ${startX} ${startY} L ${endX} ${endY}`}
          fill='transparent'
          onMouseDown={this.onPathMouseDown}
          onMouseUp={selectLink}
          stroke={selected ? primaryColor : linkColor}
          strokeWidth={lineWidth}
        />
        <rect
          fill={(selected || sourceSelected) ? darkPrimaryColor : linkColor}
          height={pinSize}
          onMouseDown={this.onSourceMouseDown}
          width={pinSize}
          x={x1}
          y={y1}
        />
        {to ? (
          <rect
            fill={(selected || targetSelected) ? darkPrimaryColor : linkColor}
            height={pinSize}
            onMouseDown={this.onTargetMouseDown}
            width={pinSize}
            x={x2}
            y={y2}
          />
        ) : null}
      </g>
    )
  }
}
