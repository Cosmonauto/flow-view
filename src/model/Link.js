import Item from './Item'
import enumerableProps from './enumerableProps'
import validate from './validate'

class FlowViewLink extends Item {
  constructor (data) {
    validate.Link(data)

    super(id)

    enumerableProps(this, data)
  }
}

export default FlowViewLink
