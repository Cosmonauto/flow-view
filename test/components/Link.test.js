import React from 'react'
import Link from 'components/Link'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const tree = renderer
    .create(<Link x1={0} y1={0} x2={10} y2={10} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
