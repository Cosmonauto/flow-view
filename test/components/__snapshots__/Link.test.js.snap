// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<g
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <path
    d="M 5 5 C 5 5, 15 15 ,15 15"
    fill="transparent"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    stroke="darkgrey"
    strokeWidth={3}
  />
  <rect
    fill="darkgrey"
    height={10}
    onMouseDown={[Function]}
    width={10}
    x={0}
    y={0}
  />
</g>
`;
